### PR TITLE
Change dict input and fix SubstitutionTransformationComponent

### PR DESCRIPTION
**Pull request by gpetretto**
There is a problem with the SubstitutionTransformationComponent, appearing also on the MP website, and this PR aims at fixing it. This involves a change to the `MPComponent.get_dict_input` method, so I am open to suggestions for a better solution.

### The problem
  
From what I have seen, the problem arises from the fact that a dict should be generated as input for the transformation and a `DataTable` is used to get the input from the interface
https://github.com/materialsproject/crystaltoolkit/blob/ea6a5a61980af4fe6f610f1665bb932a6275e9c3/crystal_toolkit/core/mpcomponent.py#L628
However, its content is not accessible through a `value`  attribute, as for the other input components, but rather a `data`. So here:
https://github.com/materialsproject/crystaltoolkit/blob/ea6a5a61980af4fe6f610f1665bb932a6275e9c3/crystal_toolkit/components/transformations/core.py#L225
the values are never fetched.

### Alternative solutions

I suppose that the best way to fix this would be to implement a proper component in the mp-react-components package to get a dict value, but I don't have the technical knowledge to implement it properly.

I also considered modifying the `MPComponent.reconstruct_kwargs_from_state` method to handle `data` properties and also fetch the `State(self.get_all_kwargs_id(), "data")` states. However it seemed that this would require a series of _ad hoc_ assumptions with the risk of breaking something else.

### This PR

I replaced the `get_dict_input`, dropping entirely the `DataTable`and using a list of pairs of input text boxes instead. It is modeled on the `get_numerical_input` method, as it is basically a grid of inputs. This allows to keep working with the `value` property. The dict is then reconstructed in the `reconstruct_kwargs_from_state`. 

This is the look of the app:
dict

Since the `get_dict_input` method is not used anywhere else in the crystal toolkit this should not break anything internally. I am not sure if the `get_dict_input` was used elsewhere, but it is not clear to me if this ever worked properly. Alternatively I can rename the method I implemented and keep the other with the `get_dict_input` name.

### CODE REVIEW

File: crystal_toolkit/components/transformations/substitution.py [hunk 1]
```diff
@@ -25,16 +25,16 @@ def options_layouts(self, state=None, structure=None):
 
         if structure and structure.is_ordered:
             species_mapping = {
-                el: None for el in [str(el) for el in structure.types_of_specie]
+                el: el for el in [str(el) for el in structure.types_of_specie]
```
Comment (janosh) [inline crystal_toolkit/components/transformations/substitution.py:28]:
```suggestion
                el: el for el in map(str, structure.types_of_specie)
```

File: crystal_toolkit/core/mpcomponent.py [hunk 2]
```diff
@@ -603,41 +602,123 @@ def get_choice_input(
...
-        default: Any | None = None,
+        default: dict | None = None,
         state: dict | None = None,
         label: str | None = None,
         help_str: str = None,
+        dict_size: int | None = None,
         key_name: str = "key",
         value_name: str = "value",
+        **kwargs,
     ):
```
Comment (janosh) [inline crystal_toolkit/core/mpcomponent.py:613]:
```suggestion
    ) -> mpc.FilterField:
```

**Comment by gpetretto** [comment]
Hi @janosh, sorry, but did you maybe have the time to look if this might be an acceptable change? Or does this also need to be approved by @mkhorton?

**Comment by janosh** [comment]
@gpetretto I think this looks great! Sorry for the delay here. I didn't merge because @mkhorton is encouraging everyone to merge their own PRs to avoid bottlenecks in dev speed from PR reviews.

File: crystal_toolkit/core/mpcomponent.py [hunk 3]
```diff
@@ -603,41 +602,120 @@ def get_choice_input(
...
+    #             {"id": "key", "name": key_name},
+    #             {"id": "value", "name": value_name},
+    #         ],
+    #         data=[{"key": k, "value": v} for k, v in default.items()],
+    #         editable=True,
+    #         persistence=False,
+    #     )
+    #
+    #     return add_label_help(dict_input, label, help_str)
+
```
Comment (janosh) [inline crystal_toolkit/core/mpcomponent.py:718]: I think we can remove the old implementation?

Comment (gpetretto) [reply crystal_toolkit/core/mpcomponent.py]: yes, I removed it.

**Comment by gpetretto** [comment]
Thanks for checking. 

Concerning the merging, I would have asked to have this PR reviewed anyway, given the changes involved, but as far as I can see I don't have the rights to merge it.

**Comment by janosh** [comment]
> I don't have the rights to merge it.

@gpetretto Ah, sorry about that.

@mkhorton I went ahead and invited Guido with write access.

**Comment by mkhorton** [comment]
Late to the thread -- @gpetretto, thanks for this PR, I appreciate the fix. This was on the to-do list but I didn't have an issue open, so I really appreciate you getting to it first 🙏

@janosh thanks for commenting + providing write access. I think all the co-authors should have write access now, but maybe I've missed someone. The PR template has been updated and I'm definitely encouraging everyone to share this codebase.

**Pull request merged**